docs/design/specs/Memory-model.md

```diff
@@ -0,0 +1,276 @@
+     - The writes performed by `volatile. cpblk` will not be observable earlier than the effects of preceeding reads and writes.
+     - No read or write that is later in the program order may be speculatively executed before the reads performed by `volatile. cpblk`
+     - `cpblk` may be implemented as a sequence of reads and writes. The granularity and mutual order of such reads and writes is unspecified.
+
+Note that volatile semantics does not by itself imply that operation is atomic or has any effect on how soon the operation is committed to the coherent memory. It only specifies the order of effects when they eventually become observable.
+
+`volatile.` and `unaligned.` IL prefixes can be combined where both are permitted.
+
+It may be possible for an optimizing compiler to prove that some data is accessible only by a single thread. In such case it is permitted to omit volatile semantics when accessing such data.
+
+* **Full-fence operations**
+  Full-fence operations have "full-fence semantics" - effects of reads and writes must be observable no later or no earlier than a full-fence operation according to their relative program order.
+  Operations with full-fence semantics:
+     - `System.Thread.MemoryBarrier`
+     - `System.Threading.Interlocked` methods
+
+## C# `volatile` feature
+One common way to introduce volatile memory accesses is by using C# `volatile` language feature. Declaring a field as `volatile` does not have any effect on how .NET runtime treats the field. The decoration works as a hint to the C# compiler itself (and compilers for other .Net languages) to emit reads and writes of such field as  reads and writes with `volatile.` prefix.
+
+## Process-wide barrier
+Process-wide barrier has full-fence semantics with an additional guarantee that each thread in the program effectively performs a full fence at arbitrary point synchronized with the process-wide barrier in such a way that effects of writes that precede both barriers are observable by memory operations that follow the barriers.
+
+The actual implementation may vary depending on the platform. For example interrupting the execution of every core in the current process' affinity mask could be a suitable implementation.
+
+## Synchronized methods
+Methods decorated with ```MethodImpl(MethodImplOptions.Synchronized)``` attribute have the same memory access semantics as if a lock is acquired at an entrance to the method and released upon leaving the method.
+
+## Data-dependent reads are ordered
+Memory ordering honors data dependency. When performing indirect reads from a location derived from a reference, it is guaranteed that reading of the data will not happen ahead of obtaining the reference. This guarantee applies to both managed references and unmanaged pointers.
+**Example:** reading a field, will not use a cached value fetched from the location of the field prior obtaining a reference to the instance.
+
+## Object assignment
```

Comment:
Is this essentially the opposite of data-dependent reads, or is it more subtle?

Reply:
Is this -any- write to the object's fields?

```
obj.field
static = obj
```
is clearly covered, but what about
```
// obj1 and obj2 (may) alias
obj1.field
static = obj2
```

Reply:
Is this code guaranteed to commit all of the updates at the final store?
```
obj1.field = obj2
obj2.field = value
static = obj1
```
(Is "reachable through the instance reference" direct or transitive?)

Reply:
Yes, it is essentially the opposite of data-dependent reads. Object writes must be emitted as releases, so that one could, for example, traverse a published linked list.

Reply:
Indirectly it means that compiler cannot reorder data-dependent writes. 
I.E. we cannot emit 

```cs
obj1.field = obj2
static = obj1
obj2.field = value
```

I do not think JIT violates this, but I wonder if this already follows from existing rules or we need to add some constraint on reordering of writes?